NeoML/include/NeoML/Dnn/Layers/DnnHeadAdapterLayer.h

```diff
@@ -22,60 +22,65 @@ limitations under the License.
 
 namespace NeoML {
 
-// Layer to pass data blobs throw common Head dnn
-// Unlike CompositeLayer, it allows to connect several layers to same internal network
+// CDnnHeadAdapterLayer passes data blobs between multiple external layers and a shared internal DNN (head)
+// Unlike CompositeLayer, it allows to connect several external layers to same head
 class NEOML_API CDnnHeadAdapterLayer final : public CBaseLayer {
 	NEOML_DNN_LAYER( CDnnHeadAdapterLayer )
 public:
 	explicit CDnnHeadAdapterLayer( IMathEngine& mathEngine, const char* name = nullptr )
-		: CBaseLayer( mathEngine, name == nullptr ? "CDnnHeadAdapterLayer" : name, /*isLearnable*/true ),
-		innerInputBlobs(),
-		head(nullptr),
-		internalDnn(nullptr),
-		blobsForBackward(0),
-		blobsForLearn(0),
-		firstAdapter(),
-		num(0)
+		: CBaseLayer( mathEngine, name == nullptr ? "CDnnHeadAdapterLayer" : name, /*isLearnable*/true )
 	{}
 
 	void Serialize(CArchive& archive) override;
 
 	// Internal shared Dnn between DnnHeadAdapters
-	void SetDnnHead(CPtr<CHeadStruct> _head);
+	void SetDnnHead(const CPtr<CDnnHead>& _head);
 
 	// Get Dnn head
-	CHeadStruct* GetDnnHead();
+	const CDnnHead* GetDnnHead() const { return head; };
 
 protected:
 	void Reshape() override;
 	void RunOnce() override;
 	void BackwardOnce() override;
 	void LearnOnce() override;
+	// It does not allocate outputBlobs in CBaseLayer in runOnce, because they are not used for inference.
+	// The outputBlob for CCompositeLayer are sinkLayer->GetBlob() of its internalDnn.
 	void AllocateOutputBlobs() override {}
+	int BlobsForBackward() const override { return blobsForBackward; }
+	int BlobsForLearn() const override { return blobsForLearn; }
 
 private:
-	// Stores the input blobs from last Inference
+	// Stores the input/output blobs from last Inference
 	CObjectArray<CDnnBlob> innerInputBlobs;
+	CObjectArray<CDnnBlob> innerOutputBlobs;
 	// Pointer to HeadLayer with inner dnn
-	CPtr<CHeadStruct> head;
-	CDnn* internalDnn; // = head->dnn
+	CPtr<CDnnHead> head = nullptr;
 	// Which of the blobs will be used during backward
-	int blobsForBackward;
+	int blobsForBackward = 0;
```

Comment:
`blobsForBackward`, `blobsForLearn`, `inputLayers`, `outputLayers`, `sourceLayers`, `sinkLayers` are unique for CDnnHead and all adapters